### PR TITLE
Refactor

### LINKED ISSUE #23: Resume the training process
When the user doesn't modify the `mode` and `target` attributes of the monitor, the process would be forced to stop.
After the early stopping:
- the user would like to resume the training process for more epoches.
- the user would like to change the random seed, the learning rate, or the optimizer and resume the process

### PR DESCRIPTION
**Pull request by peter850706**
- Transform related
  - Set compose as the factory method of `Compose`
  - `Compose` instance will still return tuple even if there is only one image.

     ```python
     img = np.array([0, 1])
     to_tensor = Compose([ToTensor()])
     img = to_tensor(img) # (old)
     img, = to_tensor(img) # (now)
     ```
  - New interface for transforms runtime kwargs and use `transformed` to specify whether the corresponding image should be transformed.
     ```python
     img1, img2, img3 = np.array([0, 1]), np.array([0, 1]), np.array([0, 1])
     to_tensor = Compose([ToTensor()])
     transforms_kwargs = {
         'ToTensor':{
             'transformed': (True, True, False)
             'dtypes': (torch.float, torch.long)
         }
     }
     img1, img2, img3 = to_tensor(img1, img2, img3, **transforms_kwargs)
     print(img1.dtype) # torch.float
     print(img2.dtype) # torch.long
     print(type(img3)) # numpy.ndarray
     ```
  - ToTensor new argument `channel_first` (default is `True`)
     If `channel_first is True`, the data format will be `CHW`, `CDHW`.
     New default behavior: infer the data type of the input images (`numpy.ndarry`).
  - New transforms
    - `MinMaxScale`
    - `Clip`
    - `Resample`
    - `RandomElasticDeform` (renamed)
- New Losses and Metrics
  - `CrossEntropyLoss`
  - `DiceLoss`
  - `TverskyLoss`
  - `Dice`
- `LossFns`, `LossWeights` and `MetricFns` are `namedtuple`, so they are indexable and iterable.
- Miscellaneous
  - The config file is copied instead of dumped.
  - Do not implicitly change `LossFns` and `MetricFns` attributes to snake case.
  - Use `LossWeights` type for user-defined weighting the computed losses (has the same usage as `LossFns` and `MetricFns`).
  - Add error messages in `Monitor` for easy debugging.
  - Add new attr `saved_dir` to `BaseTrainer` and `BasePredictor`.
  - `metric_fns` is optional now and can be a loss_fn defined in `torch.nn`.
  - Avoid replication of numpy random seed across child processes.
  - Move the `valid_freq` from `BaseTrainer` to `Monitor`.
  - Closes #23.

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -244,23 +233,16 @@ def _parse_args():
 def _get_instance(module, config, *args):
     """
     Args:
-        module (module): The python module.
-        config (Box): The config to create the class object.
+        module (MyClass): The defined module (class).
```
Comment (vic85821) [inline src/main.py:253]: The term "MyClass" is commonly used in the Python code?

File: src/model/losses.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self.smooth = smooth
+        self.square = square
+        if weight is not None:
+            weight = torch.tensor(weight, dtype=torch.float)
+            self.register_buffer('weight', weight)
+
+    def forward(self, output, target):
+        """
+        Args:
+            output (torch.Tensor) (N, C, *): The output probability.
```
Comment (vic85821) [inline src/model/losses.py]: I'm not sure whether it's necessary to consider `C=1` case.

Comment (vic85821) [reply src/model/losses.py]: If the user applies the `sigmoid` function as the final activation function.

File: src/model/losses.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        self.beta = beta
+        self.smooth = smooth
+        if weight is not None:
+            weight = torch.tensor(weight, dtype=torch.float)
+            self.register_buffer('weight', weight)
+
+    def forward(self, output, target):
+        """
+        Args:
+            output (torch.Tensor) (N, C, *): The output probability.
```
Comment (vic85821) [inline src/model/losses.py]: Same as line 51.

File: src/model/metrics.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    """The Dice score.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, output, target):
+        """
+        Args:
+            output (torch.Tensor) (N, C, *): The output probability.
```
Comment (vic85821) [inline src/model/metrics.py]: Same as `DiceLoss`.

File: src/runner/predictors/base_predictor.py [hunk 5]
```diff
@@ -14,7 +14,7 @@ class BasePredictor:
         test_dataloader (Dataloader): The testing dataloader.
         net (BaseNet): The network architecture.
         loss_fns (LossFns): The loss functions.
-        loss_weights (torch.Tensor): The corresponding weights of loss functions.
+        loss_weights (LossWeights): The corresponding weights of loss functions.
```
Comment (vic85821) [inline src/runner/predictors/base_predictor.py:18]: Do users need to weigh the loss by `loss_weights` in `test_step` by themselves?

File: src/runner/trainers/base_trainer.py [hunk 6]
```diff
@@ -21,7 +21,7 @@ class BaseTrainer:
         valid_dataloader (Dataloader): The validation dataloader.
         net (BaseNet): The network architecture.
         loss_fns (LossFns): The loss functions.
-        loss_weights (torch.Tensor): The corresponding weights of loss functions.
+        loss_weights (LossWeights): The corresponding weights of loss functions.
```
Comment (vic85821) [inline src/runner/trainers/base_trainer.py:26]: Same as the question in `base_predictor`.

File: src/runner/utils.py [hunk 7]
```diff
@@ -11,34 +11,36 @@ def __init__(self):
         self.count = 0
         self.log = None
 
-    def _init_log(self, losses, metrics=None):
+    def _init_log(self, losses=None, metrics=None):
         """Initilize the log.
```
Comment (vic85821) [inline src/runner/utils.py]: Typo.

**Review by vic85821 — COMMENTED** [review]
The code review of the `transforms.py` has not been conducted.

File: src/main.py (hunk 1 above)
Comment (peter850706) [reply src/main.py:253]: [FYI](https://docs.python.org/3/tutorial/classes.html#class-objects). I'm not very sure though.

File: src/model/losses.py (hunk 2 above)
Comment (peter850706) [reply src/model/losses.py]: Not very sure. Maybe we can mimic the [CrossEntropyLoss](https://pytorch.org/docs/stable/nn.html#crossentropyloss) in Pytorch that accepts the output logits instead of the output probability.

File: src/model/losses.py (hunk 3 above)
Comment (peter850706) [reply src/model/losses.py]: Same as [this comment](https://github.com/cmlab-mira/template/pull/25#discussion_r385621504).

File: src/model/metrics.py (hunk 4 above)
Comment (peter850706) [reply src/model/metrics.py]: Same as [this comment](https://github.com/cmlab-mira/template/pull/25#discussion_r385621504).

File: src/runner/predictors/base_predictor.py (hunk 5 above)
Comment (peter850706) [reply src/runner/predictors/base_predictor.py:18]: Yes. Besides, the users can fully control the returned keys of `_test_step` except `loss`. For example,
```python
def _test_step(self, batch):
    ...
    # The returned keys are up to you.
    my_loss_str = 'MyLoss'
    my_loss_str = 'my_loss'
    my_loss_str = 'myLoss'
    return {
        'loss': loss
        'losses':{
            my_loss_str: my_loss
        }
    }
```

File: src/runner/trainers/base_trainer.py (hunk 6 above)
Comment (peter850706) [reply src/runner/trainers/base_trainer.py:26]: Same as [this comment](https://github.com/cmlab-mira/template/pull/25#discussion_r385628206).

File: src/data/transforms.py [hunk 8]
```diff
@@ -1,169 +1,233 @@
...
+                raise TypeError('All of the transforms should be BaseTransform.')
+            self.transforms = tuple(transforms)
 
-    def __init__(self):
-        pass
+    def __call__(self, *imgs, **transforms_kwargs):
+        """
+        Args:
+            imgs (tuple of numpy.ndarray): The images to be transformed.
+            transforms_kwargs (dict): The runtime kwargs for each transforms.
```
Comment (vic85821) [inline src/data/transforms.py:43]: It can provide a simple example to demonstrate how to use the `kwargs`.

File: src/data/transforms.py [hunk 9]
```diff
@@ -1,169 +1,233 @@
...
 
-    def __call__(self, *imgs, **kwargs):
-        raise NotImplementedError
+        Returns:
+            imgs (tuple): The transformed images.
+        """
+        for transform in self.transforms:
+            kwargs = transforms_kwargs.get(transform.__class__.__name__, {})
+            transformed = kwargs.pop('transformed', tuple(True for _ in range(len(imgs))))
```
Comment (vic85821) [inline src/data/transforms.py:50]: May need to add the comments to remind the user the transforms will be applied to all images by default.

File: src/data/transforms.py [hunk 10]
```diff
@@ -1,169 +1,233 @@
...
+                raise ValueError('The number of the transformed should be the same as the images.')
+            if not all(_transformed in [True, False] for _transformed in transformed):
+                raise ValueError('All of the transformed should be either True or False.')
+            if not any(transformed):
+                raise ValueError('The transformed should not be all False.')
+
+            if all(transformed):
+                imgs = transform(*imgs, **kwargs)
+            else:
+                transformed_imgs, untransformed_imgs = self._split_imgs(imgs, transformed)
```
Comment (vic85821) [inline src/data/transforms.py:61]: We can try to use `numpy.array` here.
Like
```
_imgs = np.array(imgs)
transformed_imgs = _imgs [np.array(transform) == True]
```

File: src/data/transforms.py [hunk 11]
```diff
@@ -1,169 +1,233 @@
...
+                transformed_imgs = transform(*transformed_imgs, **kwargs)
+                imgs = self._reassemble_imgs(transformed_imgs, untransformed_imgs, transformed)
+        return imgs
 
     def __repr__(self):
```
Comment (vic85821) [inline src/data/transforms.py:66]: where is the `print function` called

File: src/data/transforms.py [hunk 12]
```diff
@@ -1,169 +1,233 @@
...
-        return self.__class__.__name__
+        format_string = self.__class__.__name__ + '('
+        for transform in self.transforms:
+            format_string += '\n'
+            format_string += f'    {transform}'
+        format_string += '\n)'
+        return format_string
 
+    @classmethod
+    def compose(cls, transforms=None):
```
Comment (vic85821) [inline src/data/transforms.py:75]: where would the `compose` be called

File: src/data/transforms.py [hunk 13]
```diff
@@ -1,169 +1,233 @@
...
+        if dtypes is None:
+            dtypes = tuple(None for _ in range(len(imgs)))
+        if len(dtypes) != len(imgs):
+            raise ValueError('The number of the dtypes should be the same as the images.')
+        imgs = tuple(torch.as_tensor(img, dtype=dtype) for img, dtype in zip(imgs, dtypes))
+        if self.channel_first:
+            if imgs[0].ndim == 3:
+                imgs = tuple(img.permute(2, 0, 1).contiguous() for img in imgs)
+            elif imgs[0].ndim == 4:
+                imgs = tuple(img.permute(3, 2, 0, 1).contiguous() for img in imgs)
```
Comment (vic85821) [inline src/data/transforms.py:180]: `elif self.channel_first == False` ...

File: src/data/transforms.py (hunk 8 above)
Comment (peter850706) [reply src/data/transforms.py:43]: Do you mean [this](https://github.com/cmlab-mira/template/pull/25#issue-378559465) should be written in the doc string?

File: src/data/transforms.py (hunk 9 above)
Comment (peter850706) [reply src/data/transforms.py:50]: I think the interface is straightforward. For example,
```python
img1, img2, img3 = np.array([0, 1]), np.array([0, 1]), np.array([0, 1])
to_tensor = Compose([ToTensor()])
img1, img2, img3 = to_tensor(img1, img2, img3)
```
The users would expect `img1, img2, img3` are transformed. That's why we need to provide the `transformed` kwarg for the users to control which image should be transfromed.

File: src/data/transforms.py (hunk 10 above)
Comment (peter850706) [reply src/data/transforms.py:61]: I have tried this in 
9f66070 but there exists some problems.
According to [np.array documentation](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html), the `dtype` will be determined as **the minimum type required to hold the objects in the sequence**. Therefore, if `imgs` are a sequence of `np.float32` and `np.int64`, `_imgs` will be transformed to a sequence of `np.float64` and `np.float64`!!

File: src/data/transforms.py (hunk 11 above)
Comment (peter850706) [reply src/data/transforms.py:66]: Sorry, I don't understand what's your concerns XD.

File: src/data/transforms.py (hunk 12 above)
Comment (peter850706) [reply src/data/transforms.py:75]: It's a factory method of `Compose`, so it can be called to create a `Compose` instance with `transforms (BoxList)`. I think when defining `MyDataset` is where it can be called, for example,
```python
class MyDataset(BaseDataset):
    def __init__(self, transforms, **kwargs):
        super().__init__(**kwargs)
        self.transforms = Compose.compose(transforms)
```

File: src/data/transforms.py (hunk 13 above)
Comment (peter850706) [reply src/data/transforms.py:180]: if `self.channel_first is False`, we do not need to do anything here.

File: src/data/transforms.py (hunk 8 above)
Comment (vic85821) [reply src/data/transforms.py:43]: Yes.

File: src/data/transforms.py (hunk 11 above)
Comment (vic85821) [reply src/data/transforms.py:66]: I mean I don't see the `print(compose)` in any code.

File: src/data/transforms.py [hunk 14]
```diff
@@ -173,29 +237,215 @@ def _normalize(img, means, stds):
...
+        Args:
+            img (numpy.ndarray): The image to be scaled.
+            mins (numpy.ndarray): The minimum values for each channel.
+            maxs (numpy.ndarray): The maximum values for each channel.
+            value_range (numpy.ndarray): The minimum and minimum value after scaling.
+
+        Returns:
+            img (numpy.ndarray): The scaled image.
+        """
+        img = img.copy()
```
Comment (vic85821) [inline src/data/transforms.py]: need `img.clip(min, max)` ?

File: src/data/transforms.py (hunk 11 above)
Comment (peter850706) [reply src/data/transforms.py:66]: I think that depends on the users, we just provide this functionality for printing some useful information ([ref](https://stackoverflow.com/questions/1436703/difference-between-str-and-repr)).

File: src/model/losses.py (hunk 2 above)
Comment (vic85821) [reply src/model/losses.py]: The same as the `BinaryCrossEntropy` (ref: https://pytorch.org/docs/stable/nn.functional.html?highlight=binary%20cross%20entropy#torch.nn.functional.binary_cross_entropy), we can implement the binary_dice_loss function.

File: src/runner/trainers/base_trainer.py (hunk 6 above)
Comment (vic85821) [reply src/runner/trainers/base_trainer.py:26]: Record in #27 .

File: src/runner/utils.py (hunk 7 above)
Comment (vic85821) [reply src/runner/utils.py]: Fixed in 05b97a7.

Comment (peter850706) [reply src/runner/utils.py]: Fix in 05b97a7.

File: src/runner/predictors/base_predictor.py (hunk 5 above)
Comment (vic85821) [reply src/runner/predictors/base_predictor.py:18]: Record in #27.

File: src/data/transforms.py (hunk 9 above)
Comment (vic85821) [reply src/data/transforms.py:50]: confirmed

File: src/data/transforms.py (hunk 8 above)
Comment (peter850706) [reply src/data/transforms.py:43]: Ok. Has metioned in #27.

File: src/data/transforms.py (hunk 12 above)
Comment (vic85821) [reply src/data/transforms.py:75]: Record in #27 .

File: src/data/transforms.py (hunk 13 above)
Comment (vic85821) [reply src/data/transforms.py:180]: confirmed.

**Comment by peter850706** [comment]
Closes #23.
